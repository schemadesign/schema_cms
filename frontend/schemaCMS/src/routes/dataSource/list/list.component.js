import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import { Card, Icons, Typography } from 'schemaUI';
import { always, anyPass, cond, equals, ifElse } from 'ramda';
import { FormattedMessage } from 'react-intl';

import { TopHeader } from '../../../shared/components/topHeader';
import {
  Container,
  DataSourceItem,
  DataSourceList,
  Error,
  ErrorsWrapper,
  Header,
  HeaderIcon,
  iconSourceStyles,
  lockIconStyles,
  lockTextStyles,
  MetaData,
  MetaDataName,
  MetaDataValue,
  MetaDataWrapper,
  titleStyles,
} from './list.styles';
import messages from './list.messages';
import extendedDayjs, { BASE_DATE_FORMAT } from '../../../shared/utils/extendedDayjs';
import { HeaderItem, HeaderList } from '../../project/list/list.styles';
import {
  FIELDS_STEP,
  INITIAL_STEP,
  STATUS_DONE,
  STATUS_DRAFT,
  STATUS_ERROR,
  STATUS_PROCESSING,
  STATUS_READY_FOR_PROCESSING,
} from '../../../modules/dataSource/dataSource.constants';
import { renderWhenTrueOtherwise } from '../../../shared/utils/rendering';
import { NavigationContainer, BackArrowButton, PlusButton } from '../../../../shared/components/navigation';

const { H1 } = Typography;
const { CsvIcon, IntersectIcon } = Icons;
const DEFAULT_VALUE = 'â€”';

export class List extends PureComponent {
  static propTypes = {
    createDataSource: PropTypes.func.isRequired,
    fetchDataSources: PropTypes.func.isRequired,
    cancelFetchListLoop: PropTypes.func.isRequired,
    dataSources: PropTypes.array.isRequired,
    intl: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        projectId: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
  };

  componentDidMount() {
    const projectId = this.props.match.params.projectId;

    this.props.fetchDataSources({ projectId });
  }

  componentWillUnmount() {
    this.props.cancelFetchListLoop();
  }

  getHeaderAndMenuConfig = () => ({
    headerTitle: this.props.intl.formatMessage(messages.title),
    headerSubtitle: this.props.intl.formatMessage(messages.subTitle),
    secondaryMenuItems: [
      {
        label: this.props.intl.formatMessage(messages.projectDetails),
        to: `/project/view/${this.props.match.params.projectId}`,
      },
    ],
  });

  getStep = ifElse(equals(STATUS_DRAFT), always(INITIAL_STEP), always(FIELDS_STEP));

  handleShowProject = () => this.props.history.push(`/project/view/${this.props.match.params.projectId}`);

  handleCreateDataSource = () => {
    const projectId = this.props.match.params.projectId;

    this.props.createDataSource({ projectId });
  };

  handleShowDataSource = ({ id, status }) => {
    if ([STATUS_PROCESSING, STATUS_READY_FOR_PROCESSING].includes(status)) {
      return;
    }

    const step = this.getStep(status);

    this.props.history.push(`/project/view/${this.props.match.params.projectId}/datasource/view/${id}/${step}`);
  };

  renderCreatedInformation = list => (
    <Header>
      <HeaderList>
        {list.map((item, index) => (
          <HeaderItem key={index}>{item}</HeaderItem>
        ))}
      </HeaderList>
      <HeaderIcon>
        <IntersectIcon />
      </HeaderIcon>
    </Header>
  );

  renderErrorMessage = () => <FormattedMessage {...messages.error} />;
  renderProcessingMessage = () => <FormattedMessage {...messages.processing} />;
  renderDraftMessage = () => <FormattedMessage {...messages.draft} />;

  renderHeader = (status, list) =>
    cond([
      [equals(STATUS_DONE), () => this.renderCreatedInformation(list)],
      [equals(STATUS_ERROR), this.renderErrorMessage],
      [anyPass([equals(STATUS_PROCESSING), equals(STATUS_READY_FOR_PROCESSING)]), this.renderProcessingMessage],
      [equals(STATUS_DRAFT), this.renderDraftMessage],
    ])(status);

  renderMetaData = ({ items, fields, filters, views }, isLock) => {
    const customIconStyles = isLock ? { ...iconSourceStyles, ...lockIconStyles } : iconSourceStyles;
    const { formatMessage } = this.props.intl;
    const list = [
      { name: formatMessage(messages.source), value: <CsvIcon customStyles={customIconStyles} /> },
      { name: formatMessage(messages.items), value: items },
      { name: formatMessage(messages.fields), value: fields },
      { name: formatMessage(messages.filters), value: filters },
      { name: formatMessage(messages.views), value: views },
    ];

    const elements = list.map(({ name, value }, index) => (
      <MetaData key={index}>
        <MetaDataName>{name}</MetaDataName>
        <MetaDataValue isLock={isLock}>{value || DEFAULT_VALUE}</MetaDataValue>
      </MetaData>
    ));

    return <MetaDataWrapper>{elements}</MetaDataWrapper>;
  };

  renderErrors = errorLog => errorLog.map((error, index) => <Error key={index}>{error}</Error>);

  renderCardErrors = errorLog => <ErrorsWrapper>{this.renderErrors(errorLog)}</ErrorsWrapper>;

  renderCardContent = ({ metaData, isLock, isError, errorLog = [] }) =>
    renderWhenTrueOtherwise(
      always(this.renderCardErrors(errorLog)),
      always(this.renderMetaData(metaData || {}, isLock))
    )(isError);

  renderItem = ({ name, created, createdBy: { firstName, lastName }, id, metaData, status, errorLog }, index) => {
    const isLock = status !== STATUS_DONE;
    const isError = status === STATUS_ERROR;
    const whenCreated = extendedDayjs(created, BASE_DATE_FORMAT).fromNow();
    const header = this.renderHeader(status, [whenCreated, `${firstName} ${lastName}`]);
    const customTitleStyles = isLock ? { ...titleStyles, ...lockTextStyles } : titleStyles;

    return (
      <DataSourceItem key={index}>
        <Card headerComponent={header}>
          <H1 customStyles={customTitleStyles} onClick={() => this.handleShowDataSource({ id, status })}>
            {name}
          </H1>
          {this.renderCardContent({ metaData, isLock, isError, errorLog })}
        </Card>
      </DataSourceItem>
    );
  };

  renderList = dataSources => <DataSourceList>{dataSources.map(this.renderItem)}</DataSourceList>;

  render() {
    const { dataSources = [] } = this.props;
    const topHeaderConfig = this.getHeaderAndMenuConfig();

    return (
      <Container>
        <Helmet title={this.props.intl.formatMessage(messages.title)} />
        <TopHeader {...topHeaderConfig} />
        {this.renderList(dataSources)}

        <NavigationContainer>
          <BackArrowButton onClick={this.handleShowProject} />
          <PlusButton onClick={this.handleCreateDataSource} />
        </NavigationContainer>
      </Container>
    );
  }
}
