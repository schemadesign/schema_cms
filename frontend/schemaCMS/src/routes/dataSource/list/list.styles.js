import styled, { css } from 'styled-components';
import { Theme } from 'schemaUI';
import { styleWhenTrue } from '../../../shared/utils/rendering';

const horizontalMargin = 20;

export const Container = styled.div`
  margin: 0 ${horizontalMargin}px 118px;
`;

export const DataSourceItem = styled.li`
  margin-top: 10px;

  &:first-child {
    margin-top: 0;
  }
`;

export const Description = styled.div`
  margin: 7px 0 14px;
  cursor: pointer;
`;

export const DataSourceList = styled.div`
  list-style: none;
  padding: 0;
  margin-top: 24px;
`;

export const titleStyles = {
  cursor: 'pointer',
  marginBottom: '20px',
};

export const MetaDataWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

export const MetaData = styled.div`
  display: flex;
  flex-direction: column;
`;

export const MetaDataName = styled.span`
  color: ${Theme.primary.label};
  font-weight: 600;
  font-size: 12px;
  margin-bottom: 10px;
`;

const lockStyles = styleWhenTrue(
  ({ isLock }) => isLock,
  css`
    color: ${Theme.primary.label};
  `
);

export const MetaDataValue = styled.span`
  font-weight: 600;
  font-size: 24px;
  line-height: 36px;

  ${lockStyles}
`;

export const Header = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;

export const HeaderIcon = styled.div`
  margin-top: -8px;
`;

export const ErrorsWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Error = styled.div``;

export const ButtonsContainer = styled.div`
  position: fixed;
  bottom: 0;
  padding: 50px 0 24px;
  background: ${Theme.primary.body};
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: calc(100% - ${2 * horizontalMargin}px);
`;

export const lockTextStyles = {
  color: Theme.primary.label,
};

export const buttonStyles = {
  backgroundColor: Theme.primary.background,
  height: '60px',
};

export const iconSourceStyles = {
  width: 40,
  height: 40,
  fill: Theme.primary.iconDark,
};

export const lockIconStyles = {
  fill: Theme.primary.label,
};
