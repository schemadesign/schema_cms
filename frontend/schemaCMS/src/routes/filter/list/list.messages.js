/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  noFields: {
    id: 'filter.list.noFields',
    defaultMessage: 'No fields found.',
  },
});
