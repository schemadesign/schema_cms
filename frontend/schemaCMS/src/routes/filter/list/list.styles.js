import styled from 'styled-components';

export const Container = styled.div``;

export const NoFields = styled.div`
  display: flex;
`;
