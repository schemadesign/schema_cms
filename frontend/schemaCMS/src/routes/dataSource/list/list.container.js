import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { hot } from 'react-hot-loader';
import { bindPromiseCreators, promisifyRoutine } from 'redux-saga-routines';
import { injectIntl } from 'react-intl';

import { compose } from 'ramda';
import { List } from './list.component';
import { DataSourceRoutines, selectDataSources } from '../../../modules/dataSource';

const mapStateToProps = createStructuredSelector({
  dataSources: selectDataSources,
});

export const mapDispatchToProps = dispatch => ({
  ...bindPromiseCreators(
    {
      createDataSource: promisifyRoutine(DataSourceRoutines.create),
      fetchDataSources: promisifyRoutine(DataSourceRoutines.fetchList),
      cancelFetchListLoop: promisifyRoutine(DataSourceRoutines.cancelFetchListLoop),
    },
    dispatch
  ),
});

export default compose(
  hot(module),
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  injectIntl,
  withRouter
)(List);
