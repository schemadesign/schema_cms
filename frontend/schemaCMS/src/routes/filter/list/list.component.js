import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { isEmpty, always, keys, map, pipe, path } from 'ramda';
import { FormattedMessage } from 'react-intl';

import { Container, NoFields, Field } from './list.styles';
import browserHistory from '../../../shared/utils/history';
import { renderWhenTrueOtherwise } from '../../../shared/utils/rendering';

import messages from './list.messages';

export class List extends PureComponent {
  static propTypes = {
    fetchFields: PropTypes.func.isRequired,
    fields: PropTypes.object.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        dataSourceId: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
  };

  async componentDidMount() {
    try {
      await this.props.fetchFields(path(['match', 'params'], this.props));
    } catch (e) {
      browserHistory.push('/');
    }
  }

  renderField = field => <Field key={field}>{field}</Field>;

  renderDataSourceFields = fields =>
    renderWhenTrueOtherwise(
      always(
        <NoFields>
          <FormattedMessage {...messages.noFields} />
        </NoFields>
      ),
      () =>
        pipe(
          keys,
          map(this.renderField)
        )(fields)
    )(isEmpty(fields));

  render() {
    const { fields } = this.props;

    return <Container>{this.renderDataSourceFields(fields)}</Container>;
  }
}
