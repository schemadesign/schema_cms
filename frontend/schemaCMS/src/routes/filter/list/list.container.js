import { connect } from 'react-redux';
import { bindPromiseCreators, promisifyRoutine } from 'redux-saga-routines';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { hot } from 'react-hot-loader';
import { compose } from 'ramda';

import { List } from './list.component';
import { DataSourceRoutines, selectFields } from '../../../modules/dataSource';

const mapStateToProps = createStructuredSelector({
  fields: selectFields,
});

export const mapDispatchToProps = dispatch =>
  bindPromiseCreators(
    {
      fetchFields: promisifyRoutine(DataSourceRoutines.fetchFields),
    },
    dispatch
  );

export default compose(
  hot(module),
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  withRouter
)(List);
