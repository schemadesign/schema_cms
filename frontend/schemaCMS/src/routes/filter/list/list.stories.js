import React from 'react';
import { storiesOf } from '@storybook/react';

import { List } from './list.component';

export const defaultProps = {
  fetchFields: Function.prototype,
  fields: {},
  match: {
    params: {
      dataSourceId: '1',
    },
  },
};

storiesOf('List', module).add('Default', () => <List {...defaultProps} />);
