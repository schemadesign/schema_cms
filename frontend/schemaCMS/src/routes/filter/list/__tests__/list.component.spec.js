import React from 'react';
import { shallow } from 'enzyme';
import { spy } from 'sinon';
import { expect } from 'chai';

import { List } from '../list.component';
import { defaultProps } from '../list.stories';

describe('List: Component', () => {
  const component = props => <List {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should call fetchFields on component mount', () => {
    const fetchFields = spy();
    render({ fetchFields });
    expect(fetchFields).to.have.been.calledWith({ dataSourceId: '1' });
  });
});
