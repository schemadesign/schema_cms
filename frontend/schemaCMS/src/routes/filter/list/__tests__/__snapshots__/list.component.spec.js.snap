// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List: Component should render correctly 1`] = `
<styled.div>
  <styled.div>
    <FormattedMessage
      defaultMessage="No fields found."
      id="filter.list.noFields"
      values={Object {}}
    />
  </styled.div>
</styled.div>
`;
